### PR TITLE
Refolding "simpl never" constants hiding fixpoints in position of argument of a destructor in tactic "simpl"

### LINKED ISSUE #19253: Unexpected unfolding of a recursive constant in `simpl`
### Description of the problem

The `simpl` tactic is supposed to refold after reduction the global recursive constants it unfolds. Here is an example where it fails to do so.

Note: this is fixed by #18591.

### Small Coq file to reproduce the bug

```coq
Arguments Nat.add : simpl never.
Eval simpl in fun n => (S n + 1) * n.
(*
Gives: fun n : nat =>
       n + (fix add (n0 m : nat) {struct n0} : nat := match n0 with 0 => m | S p => S (add p m) end) n 1 * n
Expected: fun n : nat => n + (n + 1) * n
*)
```

### Version of Coq where this bug occurs

8.4-8.19 (since the implementation of `simpl never`)

### Last version of Coq where the bug did not occur

_No response_

### PR DESCRIPTION
**Pull request by herbelin**
This fixes another failure to refolding fixpoints in `simpl`: the refolding mechanism was bypassed for "never simpl" constants used in position of argument of a destructor in tactic "simpl". Example:

```coq
Arguments Nat.add : simpl never.
Eval simpl in fun n => (S n + 1) * n.
(*
Before: fun n : nat =>
       n + (fix add (n0 m : nat) {struct n0} : nat := match n0 with 0 => m | S p => S (add p m) end) n 1 * n
After: fun n : nat => n + (n + 1) * n
*)
```

Fixes #19253 

Depends on #18576 (merged), #18577 (merged), #18580 (some commits are shared but order of merge is not important)

PS: The code passing the caches and flags starts to be a bit cumbersome, we may eventually use a record to collect the 6 components.

- [x] Added / updated **test-suite**.
- [x] Added **changelog**.

Overlays:
- uds-psl/coq-library-undecidability#224 (backwards compatible)
- math-comp/math-comp#1223 (backwards compatible)
- coq-community/fourcolor#58 (a priori backwards compatible)
- math-comp/odd-order#55 (a priori backwards compatible)

### CODE REVIEW

**Comment by herbelin** [comment]
There is an interesting phenomenon going there (in the undecidability library, originally using `move=> []`):
```coq
Require Import List.
Arguments List.app : simpl never.
Import ListNotations.
Goal forall a l l', 0 :: l = [a] ++ l' -> l = l'.
intros * [=].
(* before the PR: give l = l'
   after the PR:  give l = [] ++ l' *)
```
which one is expected?

Under the hood, this is because `injection` uses `simpl`. So, we have a similar example with only `simpl`:
```coq
Require Import List.
Arguments List.app : simpl never.
Import ListNotations.
Eval simpl in match [0] ++ [] with a :: l => l | [] => [] end.
(* before the PR: []       because it returns [] ++ [] where ++ is not refolded, so simpl never does not apply
   after the PR:  [] ++ [] because ++ is simpl never *)
```

 There are at least two issues:
- For `injection`: is it expected that `injection` calls `simpl` on the resulting subterms, rather, say, than `hnf_constr`, or `whd_simpl`, or `whd_simpl` without the `simpl never` database?
- For `simpl`: what semantics of `simpl never` do we want?
 
cc @gares

File: doc/changelog/04-tactics/18591-master+simpl-refold-in-destructor-arg-position.rst [hunk 1]
```diff
@@ -0,0 +1,5 @@
+- **Fixed:**
+  Refolding of constants marked as :g:`simpl never` in position of
+  argument of a destructor in :tacn:`simpl`
+  (`#18591 <https://github.com/coq/coq/pull/18591>`_,
+  by Hugo Herbelin).
```
Comment (jfehrle) [inline doc/changelog/04-tactics/18591-master+simpl-refold-in-destructor-arg-position.rst:5]: Could you more specific about the incorrect behavior that you fixed?  e.g. they should not have been refolded or ???

I suppose there's no bug report to refer to.  Hint: it's not too late to create one

**Comment by proux01** [comment]
@herbelin what is the status of this? is it still targetting 8.20?

**Comment by herbelin** [comment]
I'll come back to it at some time, but 8.20 is probably unrealistic. Delaying.

**Comment by herbelin** [comment]
> For `simpl`: what semantics of `simpl never` do we want, that is, once a destructor forces the evaluation of a "never" constant, should we consider that the successive recursive calls have to stop obeying "never" or not?

Implementing the current behavior would be difficult as it results from an interaction between not using `simpl never` in position of a destructor and non-head `simpl`. I see no easy way to tell that some recursive names in an a priori arbitrary positions of the term have to specifically bypass `simpl never` in the "strong" recursive calls to `simpl`. So, I'm instead considering adding an overlay.

File: doc/changelog/04-tactics/18591-master+simpl-refold-in-destructor-arg-position.rst (hunk 1 above)
Comment (herbelin) [reply doc/changelog/04-tactics/18591-master+simpl-refold-in-destructor-arg-position.rst]: OK, done. And message extended.

**Review by ppedrot — APPROVED** [review]

**Comment by ppedrot** [comment]
@coqbot merge now

**Pull request merged**